### PR TITLE
feat: Add `es‑to‑primitive`

### PR DESCRIPTION
**Pull request by ExE-Boss**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`: https://github.com/ljharb/es-to-primitive/pull/44#issuecomment-539032313
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

---

review?(@ljharb, @andrewbranch, @armanio123, @uniqueiniquity)

### CODE REVIEW

File: types/es-to-primitive/es2015.d.ts
Comment (ljharb) [inline types/es-to-primitive/es2015.d.ts]:
```suggestion
): string | number | bigint | boolean | symbol | null | undefined;
```

File: types/es-to-primitive/es5.d.ts
Comment (ljharb) [inline types/es-to-primitive/es5.d.ts]:
```suggestion
): string | number | bigint | boolean | null | undefined;
```
i think this can still produce a symbol, whether it’s the es5 algorithm or not.

File: types/es-to-primitive/es-to-primitive-tests.ts
Comment (ExE-Boss) [reply types/es-to-primitive/es-to-primitive-tests.ts]: That will cause the tests to fail on **TypeScript** &lt; 3

File: types/es-to-primitive/es2015.d.ts
Comment (ljharb) [inline types/es-to-primitive/es2015.d.ts]: not just any function, but only String or Number or Symbol.

Comment (ExE-Boss) [reply types/es-to-primitive/es2015.d.ts]: Well, the implemented algorithm doesn’t actually really care [what function is supplied](https://github.com/ljharb/es-to-primitive/blob/9c0dd264951a7cd5bbd8c5d1951843bf48351113/es2015.js#L47-L54), although [that might be incorrect](https://tc39.es/ecma262/#sec-toprimitive).

File: types/es-to-primitive/es2015.d.ts
Comment (ExE-Boss) [reply types/es-to-primitive/es2015.d.ts]: I’ve applied this in **TypeScript** 3.2 specific types.

File: types/es-to-primitive/es-to-primitive-tests.ts
Comment (ljharb) [reply types/es-to-primitive/es-to-primitive-tests.ts]: is there no way to support both, so that it’s any on v2 and unknown on v3?

File: types/es-to-primitive/es2015.d.ts
Comment (ljharb) [reply types/es-to-primitive/es2015.d.ts]: yeah, there should be an assertion. Either way the types should enforce it.

File: types/es-to-primitive/ts3.2/es2015.d.ts
Comment (ljharb) [inline types/es-to-primitive/ts3.2/es2015.d.ts]:
```suggestion
	input: unknown,
```

File: types/es-to-primitive/ts3.2/es5.d.ts
Comment (ljharb) [inline types/es-to-primitive/ts3.2/es5.d.ts]:
```suggestion
	input: unknown,
```

File: types/es-to-primitive/ts3.2/es-to-primitive-tests.ts
Comment (ljharb) [inline types/es-to-primitive/ts3.2/es-to-primitive-tests.ts]:
```suggestion
const any: unknown = void 0;
```

File: types/es-to-primitive/es-to-primitive-tests.ts
Comment (ExE-Boss) [reply types/es-to-primitive/es-to-primitive-tests.ts]: There is, but only from TS 3.1 using the `typesVersions` key in `package.json`.

Comment (ljharb) [reply types/es-to-primitive/es-to-primitive-tests.ts]: ok, so 3.1+ can still benefit from unknown?

**Comment by armanio123** [comment]
@ljharb are all comments addressed and is this ready to be merged?

File: types/es-to-primitive/.editorconfig
Comment (armanio123) [inline types/es-to-primitive/.editorconfig]: Could you please remove this file?

**Review by armanio123 — CHANGES_REQUESTED** [review]

**Comment by armanio123** [comment]
@ljharb I noticed that you're the package maintainer for this library. You could also place the `d.ts` file in your repo and ship it with the package as well if you're willing to maintain the file.

This is optional but recommended way of shipping `d.ts` files.

**Comment by ljharb** [comment]
@armanio123 in no way would i ever want to do that unless there was a programatic way to ensure that it stayed in sync with the unannotated JS; without that, I think it's *highly inadvisable* for non-TS code to ship with first-class types.

**Review by ljharb — APPROVED** [review]

File: types/es-to-primitive/.editorconfig
Comment (ExE-Boss) [reply types/es-to-primitive/.editorconfig]: If I do this, then it’s very likely that someone would change tabs to spaces under me.

Also, all my DT contributions have this file.

**Comment by ExE-Boss** [comment]
@armanio123

#39055 and #39058 have been approved by @andrewbranch because packages are still free to choose the superior tabs over the inferior spaces for indentation.

**Review by armanio123 — APPROVED** [review]
@Exe-Boss I have a discussion with the team about this, and we agreed the PR is ok for now. Nonetheless, we still strongly suggest to go for the defaults defined by the repo. I'm approving the PR.

**Pull request merged**
